File: src/main/java/com/github/penguin418/todolist/model/constant/TodoState.java
package com.github.penguin418.todolist.model.constant;

public enum TodoState {
    TODO,
    IN_PROGRESS,
    COMPLETED,
    PENDING;

    public static boolean isValidTransition(TodoState oldState, TodoState newState){
        if (oldState == null) return true;
        if (PENDING == oldState) return true;
        if (PENDING == newState) return IN_PROGRESS == oldState;
        return true;
    }
}

File: src/main/java/com/github/penguin418/todolist/model/constant/TodoState.kt
package com.github.penguin418.todolist.model.constant

enum class TodoState {
    TODO,
    IN_PROGRESS,
    COMPLETED,
    PENDING;

    companion object {
        fun isValidTransition(oldState: TodoState?, newState: TodoState): Boolean {
            if (oldState == null) return true
            if (PENDING == oldState) return true
            return if (PENDING == newState) IN_PROGRESS == oldState else true
        }
    }
}
